### PR TITLE
Guardrails: Implement response phase

### PR DESCRIPTION
**Pull request by liavweiss**
## Summary

- Implement response-side guardrail evaluation in the `ai_guardrails`
  filter, enabling NeMo to inspect upstream responses before they reach
  the client.
- Add `on_response_body` using `block_in_place` to bridge the sync
  Pingora constraint into the async `GuardProvider::evaluate` call
  (tracked in #51 for a proper async solution).
- Unify `record_verdict` to handle both request and response phases:
  request-phase blocks return a clean 403; response-phase blocks replace
  the body with a JSON error payload padded to the committed
  `Content-Length` (Pingora commits headers before `on_response_body`
  runs, so status code cannot be changed).
- Add `extract_response_messages` to parse OpenAI Chat Completion
  response format (`choices[].message`), fail-closed on unrecognized
  bodies.
- Add `fit_to_committed_length` utility that pads (space) or truncates
  the replacement body to match the original `Content-Length`, preventing
  `PrematureBodyEnd` framing errors.
- Remove the `phase.response: true` rejection guard from `from_config`
  and the `#[dead_code]` suppression on `GuardPhase::Response`.
- Add 18 new tests covering response body access, skip conditions,
  validation (fail-closed), provider verdicts (pass/block/error via
  wiremock), and `fit_to_committed_length` edge cases (equal, padding,
  truncation, `None` body).
- Update `nemo-guardrails.yaml` example to document both request and
  response phase behavior.

## Validation

- [x] Unit tests: 66 guardrails tests pass (`cargo test -p praxis-ai-filters --lib -- guardrails`)
- [x] Manual testing against real NeMo Guardrails instance (port 18000) and mock vLLM upstream
- [x] Verified response-phase blocking produces valid JSON error with correct padding
- [x] Verified request-phase behavior unchanged (403 on block, pass-through on success)
- [x] `make lint` (no new warnings)

## Checklist

- [x] I reviewed every changed line and can explain the change.
- [x] New capabilities include an example config and functional example test.
- [x] User-facing behavior and generated documentation are updated.
- [ ] Performance-sensitive changes include appropriate benchmark or load-test evidence.
- [x] Commits are signed and include a `Signed-off-by` trailer.

## Breaking changes

None. The `phase.response` config field already existed with a default of
`false`. Existing configs that omit it or set it to `false` are unaffected.

## Fix
https://github.com/praxis-proxy/ai/issues/50

### CODE REVIEW

**Comment by leseb** [comment]
@liavweiss did you address the bot comments?

**Comment by liavweiss** [comment]
@leseb, I addressed them now.
I also ran all the unit and integration tests I ran previously to make sure everything is still working as expected.

File: filters/src/guardrails/filter.rs [hunk 1]
```diff
@@ -136,41 +128,166 @@ impl HttpFilter for AiGuardrailsFilter {
...
+
+        let Some(bytes) = body.as_ref() else {
+            return Ok(FilterAction::Continue);
+        };
+
+        if bytes.is_empty() {
+            return Ok(FilterAction::Continue);
+        }
+
+        let messages = extract_response_messages(bytes)?;
```
[the remark this answers is not in the document]
Comment (liavweiss) [reply filters/src/guardrails/filter.rs]: Done.

File: filters/src/guardrails/filter.rs [hunk 2]
```diff
@@ -136,41 +128,166 @@ impl HttpFilter for AiGuardrailsFilter {
...
+        }
+    }
+
+    fn response_body_mode(&self) -> BodyMode {
+        BodyMode::StreamBuffer {
+            max_bytes: Some(DEFAULT_MAX_BODY_BYTES),
+        }
+    }
+
+    fn on_response_body(
```
[the remark this answers is not in the document]
Comment (liavweiss) [reply filters/src/guardrails/filter.rs:188]: I'll add integration coverage for both request and response phases in a dedicated follow-up PR.

File: filters/src/guardrails/filter.rs [hunk 3]
```diff
@@ -167,41 +171,171 @@ impl HttpFilter for AiGuardrailsFilter {
...
+
+        let Some(bytes) = body.as_ref() else {
+            return Ok(FilterAction::Continue);
+        };
+
+        if bytes.is_empty() {
+            return Ok(FilterAction::Continue);
+        }
+
+        if ctx.get_metadata(META_RESPONSE_SSE).is_some() {
```
Comment (christinaexyou) [inline filters/src/guardrails/filter.rs]: IIUC, for SSE, we still wait for the stream to finish (buffer) and forward along the buffered bytes. as a result, we modify streaming endpoints into buffered ones which can cause them to break. 

can we default to streaming instead and only buffer when we need the full body ? we can always switch from Stream to StreamBuffer later in the request pathway (see `merge_body_mode` in `body.rs`) but the StreamBuffer to Stream is not possible.

it seems like there's already precedent for this in `responses_to_chat_completions`. alternative fail-hard guardrails is enabled with SSE instead of forwarding along the request

File: filters/src/guardrails/filter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (christinaexyou) [reply filters/src/guardrails/filter.rs:188]: is there a reason why integration tests can't be added to this PR ? imo it isn't complete unless you test out the behavior

File: examples/configs/nemo-guardrails.yaml [hunk 4]
```diff
@@ -1,25 +1,36 @@
...
+#     success  - request forwarded to the upstream unchanged
+#     blocked  - 403 returned to the client (triggered rail names in body)
+#     error    - 500 returned to the client
 #
-# Start a local NeMo Guardrails instance before running this config
+#   Response phase:
+#     success  - response forwarded to the client unchanged
+#     blocked  - response body replaced with a JSON error payload
+#                (status remains 200 because headers are already committed)
+#     error    - 500 returned to the client
```
Comment (christinaexyou) [inline examples/configs/nemo-guardrails.yaml]: update this since we don't return a 500 on error

File: filters/src/guardrails/filter.rs [hunk 5]
```diff
@@ -133,6 +130,13 @@ impl HttpFilter for AiGuardrailsFilter {
         Ok(FilterAction::Continue)
     }
 
+    async fn on_response(&self, ctx: &mut HttpFilterContext<'_>) -> Result<FilterAction, FilterError> {
```
Comment (christinaexyou) [inline filters/src/guardrails/filter.rs:130]: instead of returning a `FilterError`, would it make more sense to replace the content body with error JSON similar to the `Block`so that the client receives the error message ?

File: filters/src/guardrails/filter.rs (hunk 3 above)
Comment (liavweiss) [reply filters/src/guardrails/filter.rs]: Thank you, I agree with you.

File: filters/src/guardrails/filter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (liavweiss) [reply filters/src/guardrails/filter.rs:188]: I missed that we have an integration test for the request, so I added an integration test for the response too. Thank you.

File: examples/configs/nemo-guardrails.yaml (hunk 4 above)
Comment (liavweiss) [reply examples/configs/nemo-guardrails.yaml]: Since I added an integration test for the response side, I will use nemo-guardrails-response.yaml config.
I have updated the correct error return there.

File: filters/src/guardrails/filter.rs (hunk 5 above)
Comment (liavweiss) [reply filters/src/guardrails/filter.rs:130]: Done.

**Comment by liavweiss** [comment]
@christinaexyou, thank you for your review! I’ve addressed all your comments.

@leseb, we’re good to go.

**Comment by szedan-rh** [comment]
Per @christinaexyou and @leseb - PR is ready be merged.
Thank you folks for the review.

**Review by szedan-rh — APPROVED** [review]

**Review by leseb — CHANGES_REQUESTED** [review]
A few comments:

* why are we bypassing streaming?
* why don't we handle auth anywhere? do we expect the guardrail endpoint to accept anonymous requests?
* since this is using SubRequestClient (it should), we will need to refactor this code once https://github.com/praxis-proxy/praxis/issues/1072 starts getting shape, now https://github.com/praxis-proxy/ai/issues/961

**Comment by liavweiss** [comment]
> A few comments:
> 
> * why are we bypassing streaming?
> * why don't we handle auth anywhere? do we expect the guardrail endpoint to accept anonymous requests?
> * since this is using SubRequestClient (it should), we will need to refactor this code once [praxis-proxy/praxis#1072](https://github.com/praxis-proxy/praxis/issues/1072) starts getting shape, now #961

1. Guardrails require the full response to evaluate intent, so per-chunk evaluation isn't viable. Buffering would break streaming behavior, so we currently skip SSE responses and evaluate only non-streaming responses. This is a known limitation we plan to revisit once (if) on_response_body becomes async (#51).

2. NeMo Guardrails doesn't require authentication by default - the endpoint is expected to be internal (sidecar / cluster-local).

3. Once #961 lands, we'll refactor the NeMo provider to align with the new SubRequestClient API.

**Comment by leseb** [comment]
> > A few comments:
> > 
> > * why are we bypassing streaming?
> > * why don't we handle auth anywhere? do we expect the guardrail endpoint to accept anonymous requests?
> > * since this is using SubRequestClient (it should), we will need to refactor this code once [praxis-proxy/praxis#1072](https://github.com/praxis-proxy/praxis/issues/1072) starts getting shape, now [feat(callout): route AI guardrail requests through filtered subrequests #961](https://github.com/praxis-proxy/ai/issues/961)
> 
> 1. Guardrails require the full response to evaluate intent, so per-chunk evaluation isn't viable. Buffering would break streaming behavior, so we currently skip SSE responses and evaluate only non-streaming responses. This is a known limitation we plan to revisit once (if) on_response_body becomes async ([Investigate async `on_response_body` support #51](https://github.com/praxis-proxy/ai/issues/51)).

makes sense

> 2. NeMo Guardrails doesn't require authentication by default - the endpoint is expected to be internal (sidecar / cluster-local).

ok 🤷🏻 

> 3. Once [feat(callout): route AI guardrail requests through filtered subrequests #961](https://github.com/praxis-proxy/ai/issues/961) lands, we'll refactor the NeMo provider to align with the new SubRequestClient API.

Thanks! Please comment on it so I assign you.

**Review by leseb — APPROVED** [review]
